DAOFoldManager and DAOReduceManager open/close time expressions

mcarcaso opened the pull request:
Allow the DAOFoldManager and DAOReduceManager to take expressions for openTime and closeTime. This will allow us to properly to monthly or quarterly reducing or folding.

mcarcaso committed:
Create the IdentityExpr that simply returns what it's given.

mcarcaso committed:
Create StartOfTimeSpan and EndOfTimeSpan glang expressions and use EndOfTimeSpan for EndOfDay.

mcarcaso committed:
Add new expressions to classes.js

mcarcaso committed:
Update the DAOFoldManager and DAOReduceManager to take openTime and closeTime expressions instead of a single periodLengthMs.

This opens the door for doing folding/reducing for non-uniform timespans like monthly and quarterly.

For backwards compatibility's sake, periodLengthMs was left on these two managers but is now just used to set the openTimeExpr/closeTimeExpr.

mcarcaso committed:
Make DAOReduceManager synchronized. This shouldn't make much difference since it should only be called in a cron but should be more correct.

adamvy commented on src/foam/nanos/analytics/DAOReduceManager.js:
```diff
@@ -42,19 +61,18 @@ foam.mlang.sink.GroupBy dataToReduce = (foam.mlang.sink.GroupBy) getSourceDAO()
-        .setOpenTime(new Date(rCloseTimeMs - periodMs))
-        .setCloseTime(new Date(rCloseTimeMs))
+    Date rCloseTime = (Date) getCloseTimeExpr().f(c.getCloseTime());
+    if ( ! reducedData.containsKey(rCloseTime) ) {
```
Can we make Candlestick a sink? Or a CandlestickSink so that we don't have to do this looping ourselves?  Would be much better if we just wrote.

```
getSourceDAO().where().select(GROUP_BY(Candlestick.KEY, CANDLESTICK()))
```

Where CANDLESTICK is a sink which creates a new candelstick on first put, and reduces into the existing one on every subsequent put.

adamvy replied:
Or maybe there's room for a "REDUCE" sink which takes a "reducable" and calls .reduce on its delegate for every put.

mcarcaso replied:
Yeah I like how you're thinking about this but I'm thinking we do that in a follow-up PR. This PR opens the door to being able to do monthly/quarterly candlesticks and we'll follow this up with better naming of our interfaces/methods for this stuff and a sink or two for folding and reducing.

adamvy commented on src/foam/nanos/analytics/DAOFoldManager.js:
```diff
@@ -20,34 +20,45 @@ foam.CLASS({
   methods: [
     {
       name: 'foldForState',
-      // TODO: use a more efficient key based lock
```
Is this TODO no longer relevant?  The method is still synchronized.

mcarcaso replied:
My bad, didn't mean to remove it. Started looking at actually addressing the TODO but stepped back because I don't have time to really get into it at the moment.

kgrgreer replied:
I removed/improved the synchronization.

mcarcaso replied:
Yup so this comment can finally go :)
Thanks Kevin!

mcarcaso committed:
Remove openTime from candlesticks

mcarcaso committed:
Update glangs

Remove the StartOfTimeSpan/StartOfDay glangs and make EndOfDay use Calendar rather than doing math on the milliseconds.

mcarcaso committed:
Undo removal of comment

mcarcaso committed:
Merge branch 'master' into CandlestickManagerExpressions

mcarcaso committed:
Merge branch 'master' into CandlestickManagerExpressions

mcarcaso committed:
Fix a couple things missed in a merge.
